File: by-python/py-boj/9012-괄호.py
def main():
    tc = int(input())
    for _ in range(tc):
        st = []
        braces = input()
        stop = False
        for b in braces:
            if b == '(':
                st.append(b)
            else:  # b == ')'일 때
                if st:  # st에 '("가 하나라도 있으면
                    st.pop()
                else:  # 없으면 )가 무의미하게 더 많으므로 NO 출력
                    print('NO')
                    stop = True
                    break

        if not stop:
            if st:
                print('NO')
            else:
                print('YES')


if __name__ == '__main__':
    main()

File: by-kotlin/kt-boj/solutions/9012-괄호.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val tc = readLine().toInt()

    repeat(tc) {
        val braces = readLine().toCharArray()
        val st = mutableListOf<Char>()
        var stop = false

        for (b in braces) {
            stop = false
            if (b == '(') st.add(b)
            else { // b == ')'일 때
                if (st.isNotEmpty()) st.removeLast() // st에 '("가 하나라도 있으면
                else {
                    println("NO")
                    stop = true // 없으면 )가 무의미하게 더 많으므로 NO 출력
                    break
                }
            }
        }

        if (!stop) {
            if (st.isNotEmpty()) println("NO")
            else println("YES")
        }
    }
}
